pkg/recommendation/device.go

```diff
@@ -0,0 +1,250 @@
+
+			}
+
+			// update nodeCapacityBlockDeviceMap for each node.
+			nodeCapacityBlockDeviceMap.update(nodeName, capacityBlockDevicesMap)
+		}
+
+		cStorPoolClusterRecommendationValue := nodeCapacityBlockDeviceMap.getDeviceRecommendtion(r.Request.Spec.PoolCapacity, r.Request.Spec.DataConfig)
+		cStorPoolClusterRecommendationValue.RequestSpec = r.Request.Spec
+		cStorPoolClusterRecommendation[kind] = cStorPoolClusterRecommendationValue
+	}
+
+	return cStorPoolClusterRecommendation
+}
+
+// nodeCapacityBlockDevice contains key with node name and value with capacityBlockDevice
+type nodeCapacityBlockDevices map[string]capacityBlockDevices
+
+// getDeviceRecommendtion returns the recommended block devices on a node with the given configuration.
+func (ncb nodeCapacityBlockDevices) getDeviceRecommendtion(poolCapacity resource.Quantity, raidConfig types.RaidGroupConfig) types.CStorPoolClusterRecommendation {
+
+	poolCapacityInt, ok := poolCapacity.AsInt64()
+	if !ok {
+		return types.CStorPoolClusterRecommendation{}
+	}
+
+	poolInstances := []types.PoolInstanceConfig{}
+
+	for nodeName, capacityBlockDevices := range ncb {
+
+		poolInstance := capacityBlockDevices.getPoolInstance(poolCapacityInt, raidConfig)
+		poolInstance.Node.Name = nodeName
+		poolInstance.Capacity = poolCapacity
+
+		poolInstances = append(poolInstances, poolInstance)
+
+	}
+
+	CStorPoolClusterRecommendation := types.CStorPoolClusterRecommendation{
+		Spec: types.CStorPoolClusterRecommendationSpec{
+			PoolInstances: poolInstances,
+		},
+	}
+
+	return CStorPoolClusterRecommendation
+}
+
+// capacityBlockDevices contains a key with capacity of block device and
+// value with all the block devices of that capacity.
+type capacityBlockDevices map[int64][]bdutil.MetaInfo
+
+func (cbd capacityBlockDevices) getPoolInstance(poolCapacityInt int64, raidConfig types.RaidGroupConfig) types.PoolInstanceConfig {
+	// To sort map storing (ascending order) keys in a seperate data structure.
+	// Note: map cannot be sorted.
+	capacityKeys := make([]int64, 0, len(cbd))
+	for capacity := range cbd {
+		capacityKeys = append(capacityKeys, capacity)
+	}
+	sort.SliceStable(capacityKeys, func(i, j int) bool {
+		return capacityKeys[i] < capacityKeys[j]
+	})
+
+	prevDataDevices := []types.Reference{}
+	for _, capacity := range capacityKeys {
+		dataDevices := []types.Reference{}
+
+		blockDevices := cbd[capacity]
+		count := int64(len(blockDevices))
+
+		if count < raidConfig.GroupDeviceCount {
+			continue
+		}
+
+		noOfRaidGroup := count / raidConfig.GroupDeviceCount
+
+		maxCapacity := noOfRaidGroup * raidConfig.GetDataDeviceCount() * capacity
+
+		// If required pool capacity is greater than the max capacity of
+		// the current block devices then skip this device.
+		if maxCapacity < poolCapacityInt {
+			continue
+		}
+
+		// Calculate the no of block devices to return to client.
+		noOfBlockDevices := (poolCapacityInt / capacity) * raidConfig.GroupDeviceCount
```

Comment:
reminder := requestedCapacity % (datadevicecount*capacity)
noofgroup = requestedCapacity/(datadevicecount*capacity)
if erminder!=0{
noofgroup +=1
}
noOfBlockDevices := noofgroup * groupdevice count

Reply:
check this logic